### PR TITLE
Dev tls cipher suite v9

### PR DESCRIPTION
**Pull request by fooinha**
@thus, thanx for your review.

New PR with style and code cleanup and review fixes. 
I also implemented, eve logging and client cipher suite in tls.log.

 Only missing:
-  LUA scripting functions.
- Read from comma separated file, ... yaml not done yet.

----

**Feature**

- Allows tls app-layer rules to match TLS handshake cypher suites
Example rules:

```
alert tls any any -> any any
(msg:"SURICATA TLS CIPHER SUITE CLIENT"; 
tls_cipher_suite:client,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,0xcca9;
sid:66666666; 
rev:1;
)
```
```
alert tls any any -> any any 
(msg:"SURICATA TLS CIPHER SUITE SERVER";
tls_cipher_suite:server,0xcca8;
sid:66666667;
rev:1;
)
```

- Custom tls logging for server cipher suite -> %C 
- Custom tls logging for client cipher suite -> %c
- Eve logging for tls cipher suites 
- Load list from configuration file (tls.cipher-suites.config) configured at suricata.yaml
   - Fallback to default values if configuration file load is not done.

- Documentation

Replaces #2657 / #2671 / #2685

fyi, @thus

### CODE REVIEW

File: src/app-layer-ssl.c [hunk 1]
```diff
@@ -287,11 +289,46 @@ static int TLSDecodeHandshakeHello(SSLState *ssl_state, uint8_t *input,
...
+            if (!ssl_state->curr_connp->cipher_suites) {
+                SCLogError(SC_ERR_MEM_ALLOC, "Failed to allocate client cipher suites.");
+            } else {
+                memcpy(ssl_state->curr_connp->cipher_suites, input, cipher_suites_length);
+            }
+            input += cipher_suites_length;
+        }
+    }
+    if (ssl_state->curr_connp->handshake_type == SSLV3_HS_SERVER_HELLO) {
+        ssl_state->curr_connp->cipher_suites = SCMalloc(sizeof(uint16_t));
```
Comment (inliniac) [inline src/app-layer-ssl.c]: I've seen a memleak happen here:
<pre>
==2348==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 2 byte(s) in 1 object(s) allocated from:
    #0 0x4c2e9b in __interceptor_malloc (/home/victor/qa/buildbot/donkey/z600fuzz/Private/src/suricata+0x4c2e9b)
    #1 0x837079 in TLSDecodeHandshakeHello /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/app-layer-ssl.c:321:48
    #2 0x82fdd3 in SSLv3ParseHandshakeType /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/app-layer-ssl.c:471:18
    #3 0x82ce5d in SSLv3ParseHandshakeProtocol /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/app-layer-ssl.c:682:14
    #4 0x828282 in SSLv3Decode /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/app-layer-ssl.c:1316:22
    #5 0x8201b6 in SSLDecode /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/app-layer-ssl.c:1484:30
    #6 0x819cf7 in SSLParseServerRecord /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/app-layer-ssl.c:1576:12
    #7 0x7a2463 in AppLayerParserParse /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/app-layer-parser.c:1051:13
    #8 0x53887e in TCPProtoDetect /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/app-layer.c:421:17
    #9 0x5353b1 in AppLayerHandleTCPData /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/app-layer.c:561:13
    #10 0x117bd20 in ReassembleUpdateAppLayer /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/stream-tcp-reassemble.c:1030:13
    #11 0x1178ec5 in StreamTcpReassembleAppLayer /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/stream-tcp-reassemble.c:1102:12
    #12 0x1188651 in StreamTcpReassembleHandleSegmentUpdateACK /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/stream-tcp-reassemble.c:1649:9
    #13 0x1187ac2 in StreamTcpReassembleHandleSegment /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/stream-tcp-reassemble.c:1688:9
    #14 0x114e4b1 in HandleEstablishedPacketToServer /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/stream-tcp.c:2096:9
    #15 0x10ddc9f in StreamTcpPacketStateEstablished /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/stream-tcp.c:2474:13
    #16 0x10b545e in StreamTcpPacket /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/stream-tcp.c:4534:20
    #17 0x11202c9 in StreamTcp /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/stream-tcp.c:4898:11
    #18 0xde6c01 in FlowWorker /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/flow-worker.c:215:9
    #19 0x11d6014 in TmThreadsSlotVarRun /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/tm-threads.c:128:17
    #20 0x11f63d1 in TmThreadsSlotVar /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/tm-threads.c:585:17
    #21 0x7fac05e94183 in start_thread /build/eglibc-MjiXCM/eglibc-2.19/nptl/pthread_create.c:312
</pre>

This happens in fuzzer generated traffic. I guess there are 2 possible cases:
1. memory is not freed at flow end (seems unlikely)
2. memory allocation is overwritten. This is what I suspect. Can this function be called multiple times if the traffic is weird enough?

Comment (fooinha) [inline src/app-layer-ssl.c]: I guess that we can check if cipher suites are already present and allocated in state struct.

"Double" handshake?

Comment (fooinha) [inline src/app-layer-ssl.c]: Applied a if check and pushed commit.
@inliniac , do you want a new PR?

**Comment by inliniac** [comment]
Again, please don't update exiting PRs. It's very confusing.

**Pull request closed without merging**